src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -108,89 +141,138 @@ void IRunningDocumentTableEventListener.OnRenameDocument(string newMoniker, stri
             {
             }
 
-            public static async Task<OpenFileTracker> CreateAsync(VisualStudioWorkspaceImpl workspace, IAsyncServiceProvider asyncServiceProvider)
+            public static async Task<OpenFileTracker> CreateAsync(
+                VisualStudioWorkspaceImpl workspace,
+                IThreadingContext threadingContext,
+                IAsyncServiceProvider asyncServiceProvider)
             {
                 var runningDocumentTable = (IVsRunningDocumentTable?)await asyncServiceProvider.GetServiceAsync(typeof(SVsRunningDocumentTable)).ConfigureAwait(true);
                 Assumes.Present(runningDocumentTable);
 
                 var componentModel = (IComponentModel?)await asyncServiceProvider.GetServiceAsync(typeof(SComponentModel)).ConfigureAwait(true);
                 Assumes.Present(componentModel);
 
-                return new OpenFileTracker(workspace, runningDocumentTable, componentModel);
+                return new OpenFileTracker(workspace, threadingContext, runningDocumentTable, componentModel);
             }
 
             private void TryOpeningDocumentsForMoniker(string moniker, ITextBuffer textBuffer, IVsHierarchy? hierarchy)
             {
                 _foregroundAffinitization.AssertIsForeground();
 
-                _workspace.ApplyChangeToWorkspace(w =>
-                {
-                    var documentIds = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(moniker);
-                    if (documentIds.IsDefaultOrEmpty)
-                    {
-                        return;
-                    }
+                // First clear off any existing IVsHierarchies we are watching. Any ones that still matter we will
+                // resubscribe to. We could be fancy and diff, but the cost is probably negligible.  Then watch and
+                // subscribe to this hierarchy and any related ones.
+                UnsubscribeFromWatchedHierarchies(moniker);
+                WatchAndSubscribeToHierarchies(moniker, hierarchy);
```

Comment:
important, all the hierarchy watch/subscribe/unsubscribe is still done synchronously on the UI thread.  This was needed or else things def broke :)

neither of these touch the workspace though.